skimage/registration/_affine.py

```diff
@@ -0,0 +1,201 @@
+
+__all__ = ['affine']
+
+
+def _parameter_vector_to_matrix(parameter_vector):
+    """Convert m optimization parameters to a (n+1, n+1) transformation matrix.
+
+    By default (the case of this function), the parameter vector is taken to
+    be the first n rows of the affine transformation matrix in homogeneous
+    coordinate space.
+
+    Parameters
+    ----------
+    parameter_vector : (ndim*(ndim+1)) array
+        A vector of M = N * (N+1) parameters.
+
+    Returns
+    -------
+    matrix : (ndim+1, ndim+1) array
+        A transformation matrix used to affine-map coordinates in an
+        ``ndim``-dimensional space.
+    """
+    m = parameter_vector.shape[0]
+    ndim = int((np.sqrt(4*m + 1) - 1) / 2)
+    top_matrix = np.reshape(parameter_vector, (ndim, ndim+1))
+    bottom_row = np.array([[0] * ndim + [1]])
+    return np.concatenate((top_matrix, bottom_row), axis=0)
+
+
+def cost_nmi(image0, image1, *, bins=100):
+    """Negative of the normalized mutual information.
+
+    See :func:`skimage.metrics.normalized_mutual_information` for more info.
+
+    Parameters
+    ----------
+    image0, image1 : array
+        The images to be compared. They should have the same shape.
+    bins : int or sequence of int, optional
+        The granularity of the histogram with which to compare the images.
+        If it's a sequence, each number is the number of bins for that image.
+
+    Returns
+    -------
+    cnmi : float
+        The negative of the normalized mutual information between ``image0``
+        and ``image1``.
+    """
+    return -normalized_mutual_information(image0, image1, bins=bins)
+
+
+def _param_cost(reference_image, moving_image, parameter_vector, *,
+                vector_to_matrix, cost, multichannel):
+    transformation = vector_to_matrix(parameter_vector)
+    if not multichannel:
+        transformed = ndi.affine_transform(moving_image, transformation,
+                                           order=1)
+    else:
+        transformed = np.zeros_like(moving_image)
+        for ch in range(moving_image.shape[-1]):
+            ndi.affine_transform(moving_image[..., ch], transformation,
+                                 order=1, output=transformed[..., ch])
+    return cost(reference_image, transformed)
+
+
+def affine(reference_image, moving_image, *,
+           cost=cost_nmi, initial_vector=None,
+           translation_indices=None, vector_to_matrix=None,
+           pyramid_scale=2, pyramid_minimum_size=8, multichannel=False,
+           inverse=True, level_callback=None, method='Powell',
+           **kwargs):
+    """Find a transformation matrix to register a moving image to a reference.
+
+    Parameters
+    ----------
+    reference_image : ndarray
+        A reference image to compare against the target.
+    moving_image : ndarray
+        Our target for registration. Transforming this image using the
+        returned matrix aligns it with the reference.
+    cost : function, optional
```

Comment:
Would [function or strings] be easier to understand for the user, so you can list the options as easily-readable strings?

Reply:
well, strings means we need to start providing a few... But, possibly!